### DISCUSSION TITLE
Support more than 4 controllers through Sunshine+Moonlight

### DISCUSSION THREAD

**Discussion by kimtore** [LizardByte/community#178] [Sunshine] [answered]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your issue described in the documentation?

- [X] I have read the documentation

### Is your issue present in the nightly release?

- [X] This issue is present in the nightly release

### Describe the Bug

Attempting to connect more than 4 gamepads to the client system running Moonlight yields unexpected behavior. The fifth gamepad will not be registered on the host system running Sunshine. Also, the fifth controller might displace another controller that was already connected.

Plugging the fifth controller directly into the host system works as expected.

### Expected Behavior

Support more than 4 controllers simultaneously.

### Host Operating System

Linux

### Operating System Version

Arch Linux

### Architecture

64 bit

### Sunshine commit or version

0.17.0

### Package

Linux - AUR

### GPU Type

AMD

### GPU Model

RX 5700 XT

### GPU Driver/Mesa Version

22.3.4

### Capture Method (Linux Only)

KMS

**Comment by kimtore** [comment]
Does anyone know what it would take to get this working? Am I doing something wrong? Is it a limitation of the protocol, or Moonlight, or Sunshine, or both?

**Comment by triwooder** [comment]
Hi, is your problem resolved? I encountered the same issue, and I only have 3 controllers that can be used simultaneously.

**Comment by cgutman** [comment, marked as answer]
The 4 controller limit is enforced on the Moonlight side to match GFE's limit. The theoretical protocol limit is 16 per client (number of bits in the active gamepad mask field).

I have updated most Moonlight clients to support up to 16 controllers with Sunshine recently, but the changes haven't made it into Moonlight releases yet.

**Reply by ASKUTKRIMSON** [reply]
Or please, add a more controllers support in new releases 🙏🙏🙏

**Reply by triwooder** [reply]
Hi! According to what you said, the next release version of Moonlight will support up to 16 game controllers, it sounds amazing. Does this also mean that Moonlight is about to have a new release version? It seems like it hasn't been updated for over eight months. 

I have another question too. As far as I know, the XInput protocol only supports up to four simultaneous game controllers. If I connect more controllers using Moonlight, what will happen with the extra controllers?

**Reply by ReenigneArcher** [reply]
Our last release was less than two months ago.

**Reply by triwooder** [reply]
Sorry, my expression may be ambiguous. What I mean is that the Moonlight PC release hasn't been update for long time, not sunshine.

**Comment by ASKUTKRIMSON** [comment]
@cgutman Please tell me how to increase the number of gamepads to at least 5, I really need it

**Reply by cgutman** [reply]
What platform are you running Moonlight on?

**Reply by ASKUTKRIMSON** [reply]
Pc for hosting and pc for streaming

**Reply by ASKUTKRIMSON** [reply]
Mean pc

**Reply by ASKUTKRIMSON** [reply]
@cgutman

**Reply by cgutman** [reply]
You can try the latest portable build of Moonlight for Windows: https://ci.appveyor.com/api/buildjobs/86obf0nwmskllw7l/artifacts/MoonlightPortable-x64-r2084.zip

**Reply by ASKUTKRIMSON** [reply]
downloading from this site will take me 10 hours to download, is there another site?

**Reply by ASKUTKRIMSON** [reply]
@cgutman Wtf, What does "You can try" mean, you can help me, but instead you throw me a link to a build in which it is not known whether this error has been fixed

**Reply by kimtore** [reply]
@ASKUTKRIMSON
Watch your tone. You are speaking with open source developers who give you this software free of charge. This is no place to barge in and demand answers and solutions. A little patience will get you a long way.

**Reply by jimykid86** [reply]
Hi, I am trying to get the latest nightly build or dev build of the moonlight.  I am trying to get the DS4 motion sensor to work.  Thanks.

PS: the link above that cgutman left didn't work for me for some reason.

**Reply by cgutman** [reply]
The link probably expired. Here is a link to the whole nightly pipeline - https://ci.appveyor.com/project/cgutman/moonlight-qt/history

Click on the top build, then pick your platform (Windows is the first one), and finally click on the Artifacts tab to see the downloads.

**Reply by jimykid86** [reply]
Thank you so much cgutman!  The DS4's gyro works great with the latest nightly build of sunshine on windows 11.  I was wondering if there is a ios version of moonlight r2119.  If not, how might I be able to convert that file for the ios on iPhone.  Thanks.

**Comment by ASKUTKRIMSON** [comment]
I've been trying to fix this problem for a week now

**Comment by KuleRucket** [comment]
> 10 hours to download, is there another site

If it takes you 10 hours to download a 24MB file, the issue is with you or your ISP not the download site. Although, you don't sound very amicable so I would be surprised if you get any further help.

**Comment by Darfeikus** [comment]
Was this feature eventually merged into the main build?

**Reply by ReenigneArcher** [reply]
Sunshine and official Moonlight clients should all support 16 controllers now.

Third-party moonlight clients may or may not support this.
